command/agentproxyshared/cache/lease_cache.go

```diff
@@ -440,18 +551,87 @@ func (c *LeaseCache) Send(ctx context.Context, req *SendRequest) (*SendResponse,
 	index.RequestToken = req.Token
 	index.RequestHeader = req.Request.Header
 
+	if index.Type != cacheboltdb.StaticSecretType {
+		// Store the index in the cache
+		c.logger.Debug("storing response into the cache", "method", req.Request.Method, "path", req.Request.URL.Path)
+		err = c.Set(ctx, index)
+		if err != nil {
+			c.logger.Error("failed to cache the proxied response", "error", err)
+			return nil, err
+		}
+
+		// Start renewing the secret in the response
+		go c.startRenewing(renewCtx, index, req, secret)
+	}
+
+	return resp, nil
+}
+
+func (c *LeaseCache) cacheStaticSecret(ctx context.Context, req *SendRequest, resp *SendResponse, index *cachememdb.Index) error {
+	// We must hold a lock for the index while it's being updated.
+	// We prepend "index/" to this lock so that it's distinct to the lock
+	// being held for inflight requests.
+	// We keep the two locking mechanisms distinct, so that it's only writes
+	// that have to be serial.
+	lock := locksutil.LockForKey(c.idLocks, "index/"+index.ID)
```

Comment:
Why not simply add a lock to Index?  That would allow for greater concurrency, since there's no reason why we need to block everyone else using the same lock that's not at this line.

Reply:
Hmmm, good point! That's a much better idea. I'll rework

Reply:
Updated!